storage/innobase/log/log0recv.cc

```diff
@@ -3842,17 +3844,56 @@ bool meb_scan_log_recs(
 #endif
         auto data_len = block_header.m_data_len;
 
-    if (scanned_lsn + data_len > recv_sys->scanned_lsn &&
-        recv_sys->scanned_epoch_no > 0 &&
-        !log_block_epoch_no_is_valid(block_header.m_epoch_no,
-                                     recv_sys->scanned_epoch_no)) {
-      /* Garbage from a log buffer flush which was made
-      before the most recent database recovery */
+#ifdef XTRABACKUP
+    /**
+     * 8.0.30 records LOG_BLOCK_EPOCH_NO at 8th by of log block header
+     * while pre-8.0.30 we recorded LOG_BLOCK_CHECKPOINT_NO
+     */
+    if (scanned_lsn + data_len > recv_sys->scanned_lsn) {
+      if (xtrabackup_original_log_format >= Log_format::VERSION_8_0_30 &&
+          recv_sys->scanned_epoch_no > 0 &&
+          !log_block_epoch_no_is_valid(block_header.m_epoch_no,
+                                       recv_sys->scanned_epoch_no)) {
+        /* Garbage from a log buffer flush which was made
+        before the most recent database recovery */
 
-      finished = true;
+        finished = true;
 
-      break;
+        break;
+      } else {
+        /** block_header.m_epoch_no = log_block_get_checkpoint_no
+         * recv_sys->scanned_epoch_no = recv_sys->scanned_checkpoint_no
+         *  log_block_get_checkpoint_no(log_block) <
+            recv_sys->scanned_checkpoint_no &&
+        (recv_sys->scanned_checkpoint_no -
+             log_block_get_checkpoint_no(log_block) >
+         0x80000000UL)
+         */
+        if (block_header.m_epoch_no < recv_sys->scanned_epoch_no &&
+            (recv_sys->scanned_epoch_no - block_header.m_epoch_no >
+             0x80000000UL)) {
+          /* Garbage from a log buffer flush which was made
+          before the most recent database recovery */
+
+          finished = true;
+
+          break;
+        }
+      }
     }
+#else
+        if (scanned_lsn + data_len > recv_sys->scanned_lsn &&
+            recv_sys->scanned_epoch_no > 0 &&
+            !log_block_epoch_no_is_valid(block_header.m_epoch_no,
+                                         recv_sys->scanned_epoch_no)) {
```

Comment:
During prepare, should we also deal with the flush_bit?
Our backup parser uses 80.29 parser which deals with flush_bit. 

Without the flush_bit mask, we read the wrong header_no wrong and we stop early.

Reply:
I mean log_data_block_header_deserialize() should consider FLUSH_BIT_MASK and get proper header_no

Reply:
Already done.